llvm/lib/Target/EraVM/EraVMOptimizeSelectPreRA.cpp

```diff
@@ -297,58 +301,93 @@ bool EraVMOptimizeSelectPreRA::tryFoldSelectZero(MachineBasicBlock &MBB) {
-         !InAddRegs.count(EraVM::in1Iterator(UseMI)->getReg())))
+        (!InRegs.count(EraVM::in0Iterator(UseMI)->getReg()) &&
+         !InRegs.count(EraVM::in1Iterator(UseMI)->getReg())))
       continue;
 
-    bool OutAddIsIn1Use = EraVM::out0Iterator(MI)->getReg() ==
-                          EraVM::in1Iterator(UseMI)->getReg();
+    bool OutIsIn1Use = OutReg == EraVM::in1Iterator(UseMI)->getReg();
     auto CC = getImmOrCImm(*EraVM::ccIterator(UseMI));
 
     // The COND_OF is overflow LT which hasn't reversal version, so we don't
-    //  attempt to inverse it.
-    if (OutAddIsIn1Use && CC == EraVMCC::COND_OF)
+    // attempt to inverse it.
+    if (OutIsIn1Use && CC == EraVMCC::COND_OF)
       continue;
 
-    auto CCNewMI = OutAddIsIn1Use ? InverseCond[CC] : CC;
-    Register TieReg = OutAddIsIn1Use ? EraVM::in0Iterator(UseMI)->getReg()
-                                     : EraVM::in1Iterator(UseMI)->getReg();
-
-    [[maybe_unused]] auto *NewMI = getFoldedInst(
-        UseMI, MI, EraVM::in0Range(MI), EraVM::in1Iterator(MI)->getReg(),
-        MI.getOpcode(), CCNewMI, TieReg);
-    LLVM_DEBUG(dbgs() << "== Folding add:"; MI.dump();
-               dbgs() << "       and use:"; UseMI.dump();
-               dbgs() << "          into:"; NewMI->dump(););
+    // To fold the instruction into select, we need to update the original
+    // instruction with the definition of select instruction, to update
+    // conditional code, to add implicit flags and to tie the registers.
+    auto &MF = *MBB.getParent();
+    auto *NewMI = MF.CloneMachineInstr(&MI);
+
+    // Place the new instruction right before the use instruction.
+    UseMI.getParent()->insert(UseMI, NewMI);
+
+    // Go through the defs of the new instruction and find the one that
+    // needs to be updated.
+    int DefIdxToTie = -1;
+    for (const auto &[Idx, DefMO] : enumerate(NewMI->defs())) {
+      if (DefMO.getReg() != OutReg)
+        continue;
+
+      // Change the def register to the def register of the select instruction,
+      // and remember the index to tie the operands.
+      DefMO.setReg(UseMI.getOperand(0).getReg());
+      DefIdxToTie = Idx;
+      break;
+    }
+    assert(DefIdxToTie != -1 && "Didn't find the def register to tie");
+
+    auto CCNewMI = OutIsIn1Use ? InverseCond[CC] : CC;
+    Register TieReg = OutIsIn1Use ? EraVM::in0Iterator(UseMI)->getReg()
+                                  : EraVM::in1Iterator(UseMI)->getReg();
+    EraVM::ccIterator(*NewMI)->ChangeToImmediate(CCNewMI);
+    MachineInstrBuilder(MF, NewMI)
+        .addReg(TieReg, RegState::Implicit)
+        .addReg(EraVM::Flags, RegState::Implicit);
+
+    // Add tie to ensure those two operands will get same reg
+    // after RA pass. This is the key to make transformation in this pass
+    // correct.
+    NewMI->tieOperands(DefIdxToTie, NewMI->getNumOperands() - 2);
+
+    LLVM_DEBUG(dbgs() << "== Folding arith or bitwise:"; MI.dump();
```

Comment:
nit: one can infer the class of instruction from MI.dump(), we don't need to duplicate.

Reply:
Done.